src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -298,16 +332,39 @@ public String toCheckpoint(EntityModel model) {
         });
     }
 
-    /**
-     * Deletes the model checkpoint for the id.
-     *
-     * @deprecated use deleteModelCheckpoint with listener instead
-     *
-     * @param modelId ID of the model checkpoint
-     */
-    @Deprecated
-    public void deleteModelCheckpoint(String modelId) {
-        clientUtil.<DeleteRequest, DeleteResponse>timedRequest(new DeleteRequest(indexName, modelId), logger, client::delete);
+    private String rcfModelToCheckpoint(RandomCutForest model) {
+        LinkedBuffer borrowedBuffer = null;
+        try {
+            borrowedBuffer = deserializeRCFBufferPool.borrowObject();
+            try {
+                return rcfModelToCheckpoint(model, borrowedBuffer);
+            } catch (Exception e) {
+                if (borrowedBuffer != null) {
+                    deserializeRCFBufferPool.invalidateObject(borrowedBuffer);
+                    borrowedBuffer = null;
+                }
+            } finally {
+                if (borrowedBuffer != null) {
+                    deserializeRCFBufferPool.returnObject(borrowedBuffer);
+                }
+            }
+        } catch (Exception e) {
+            logger.error("Failed to borrow an buffer from object pool", e);
+            return rcfModelToCheckpoint(model, null);
+        }
+        return null;
+    }
+
+    String rcfModelToCheckpoint(RandomCutForest model, LinkedBuffer buffer) {
+        final LinkedBuffer deserializationBuffer = buffer == null ? LinkedBuffer.allocate(deserializeRCFBufferSize) : buffer;
+        try {
+            RandomCutForestState state = mapper.toState(model);
```

Comment:
Did you want to set mapper.setSaveExecutorContextEnabled(true);

What about mapper.setSaveTreeStateEnabled(true)? Saving the trees would speed up run time dramatically.

Reply:
Maybe the initialization controls these?

Reply:
Yes, mapper is a singleton and I initiated here: https://github.com/kaituo/anomaly-detection-1/blob/compactRCF2/src/main/java/org/opensearch/ad/AnomalyDetectorPlugin.java#L361

Reply:
Ok. Do you want to save the trees? Saving the trees would be 10x faster (but some more space). Maybe have a config option so that we can decide on the future/case-by-case?

Reply:
Yes, let me add "mapper.setSaveTreeStateEnabled(true)".  That's the only change I need to make, right?

Reply:
yes.

Reply:
Also consider adding "mapper.setPartialTreeStateEnabled(true);" this will save space by 20% but serialization/de time will increase by 20%. I think we are more space constrained for HCAD at the moment (If not true, then do not add this).

Reply:
will change.  Disk will be become a bottleneck for 1 million entity.